### DISCUSSION TITLE
Defining custom class as argument to @asset metadata

### DISCUSSION THREAD

**Discussion by cody-scott** [dagster-io/dagster#21009] [Q&A]
Hi, I have a class that implements `__str__` and `__repr__` and i would like to show this as part of the asset decorator in my asset catalog.

basically i want this

```python
class MyClass:
    def __str__() -> str:
        return "string"

    def __repr__() -> str:
        return "repr"

@asset(metadata={'custom': MyClass()})
def xyz():
    ...
```

to show properly in the UI. 

right now it just shows as `[MyClass] (unserializable)`

I make use of the metadata in my custom IO manager, but also want to surface this then to the users to see what the contents of the class is. In this case its conforming my SQL types to load, so i want to make it visible the columns + types.

**Comment by vladidobro** [comment]
Hi,

this is also very important to me, can there be any protocol or something so that we can at least render `__repr__`?
I think that might be queite easy and without any bad side effects?

**Comment by bosprimigenious** [comment]
Implementing `__str__` or `__repr__` alone does not make an arbitrary object render as text in Dagster metadata. For text-only display, explicitly convert it, for example:

```python
"custom_display": dg.MetadataValue.text(str(obj))
```

However, since your use case is displaying column names and SQL types, `TableSchema` is a better fit:

```python
import dagster as dg

schema = dg.TableSchema(
    columns=[
        dg.TableColumn(name="id", type="INTEGER"),
        dg.TableColumn(name="name", type="TEXT"),
    ]
)

@dg.asset(
    metadata={
        "dagster/column_schema": dg.MetadataValue.table_schema(schema),
    }
)
def xyz():
    return [{"id": 1, "name": "Alice"}]
```

In your custom IO manager's `handle_output`, you can read this definition metadata:

```python
schema = context.definition_metadata["dagster/column_schema"].schema

for column in schema.columns:
    column_name = column.name
    sql_type_name = column.type
    # Map these to the types expected by your database library.
```

This requires adapting the IO manager if it currently expects a `MyClass` instance. If that object contains additional behavior or database-specific type objects, keep those in the IO manager/resource and expose a separate serializable representation as metadata. Converting it to text does not preserve the original object.

A local test with Dagster 1.11.16 verified metadata serialization and access from `handle_output` during asset materialization. The database-specific loading logic and UI rendering were not tested.

Docs: [Asset metadata and table schemas](https://docs.dagster.io/guides/build/assets/metadata-and-tags).

**Reply by vladidobro** [reply]
Thank you,
yes, we can **convert** to dagster metadata types, but that loses our own object.
I actually want my IO Manager to expect MyClass, **without having to marshall it back and forth through a dagster metadata type.**
For example, my concrete usecase is something like
```
import sqlalchemy as sa
import dagster as dg

metadata = sa.MetaData()
mytable = sa.Table(
    'mytable',
    metadata,
    ... # columns
)

@dg.asset(metadata={'table': mytable})
def savedata():
    return pd.DataFrame(...)

class TableIOManager(ConfigurableIOManager):
    engine: dg.ResourceDependency[sa.Engine]
    def handle_output(self, context, obj):
        table = context.metadata['table']
        table.create(engine, if_not_exists=True)
        obj.to_sql(str(table), engine, if_exists='append')
```

The point is that I want to keep my real objects which hold valuable information.
